src/main/kotlin/se/uulm/snowballr/backend/service/UserService.kt

```diff
@@ -98,77 +108,51 @@ class UserService(
-
     override suspend fun getUserById(request: Base.Id): GrpcUser = withUser(userRepo) { currentUser ->
         val targetUserId = parseUUID(request.id, EntityType.USER)
 
-        verifyUserAccess(currentUser, targetUserId, IdentifierType.ID)
+        isAllowedToReadUser(projectMemberRepo).checkFor(currentUser, targetUserId)
 
         val isRequestedUser = currentUser.id == targetUserId
 
         // Don't re-request the user if it is the current user itself
-        val result =
+        val targetUser =
             if (isRequestedUser) {
                 currentUser
             } else {
                 userRepo.getUserById(targetUserId).getOrThrow()
             }
 
-        // Only active or active unconfirmed users can be retrieved
-        if (result.status != UserStatus.USER_STATUS_ACTIVE &&
-            result.status != UserStatus.USER_STATUS_ACTIVE_UNCONFIRMED
-        ) {
-            throw NotFoundException(EntityType.USER, request.id)
-        }
+        // Only active or active unconfirmed users can be retrieved if the requester is not a server admin
+        isServerAdmin.forTarget<User>()
+            .orElse(isTargetUserActive)
+            .orElseThrow(UserNotFoundException(request.id))
+            .checkFor(currentUser, targetUser)
 
-        result.toGrpcUser()
+        targetUser.toGrpcUser()
     }
 
     override suspend fun getUserByEmail(request: Base.Email): GrpcUser = withUser(userRepo) { currentUser ->
         // We have to request the user first to get the ID for the access checks
         val targetUser = userRepo.getUserByEmail(request.email).getOrThrow()
 
-        verifyUserAccess(currentUser, targetUser.id, IdentifierType.EMAIL)
-
-        // Only active or active unconfirmed users can be retrieved
-        if (targetUser.status != UserStatus.USER_STATUS_ACTIVE &&
-            targetUser.status != UserStatus.USER_STATUS_ACTIVE_UNCONFIRMED
-        ) {
-            throw NotFoundException(EntityType.USER, request.email, identifierType = IdentifierType.EMAIL)
-        }
+        isAllowedToReadUser(projectMemberRepo, IdentifierType.EMAIL)
+            .forProperty(User::id)
+            // Only active or active unconfirmed users can be retrieved if the requester is not a server admin
+            .andAlso(
+                isServerAdmin.forTarget<User>()
+                    .orElse(isTargetUserActive)
+                    .orElseThrow(UserNotFoundException(request.email, IdentifierType.EMAIL)),
+            )
+            .checkFor(currentUser, targetUser)
 
         targetUser.toGrpcUser()
     }
 
     override suspend fun getAllUsers(): GrpcUser.List = withUser(userRepo) { currentUser ->
-        verifyServerAdminRole(currentUser) { UnauthorizedException.All(EntityType.USER, AccessType.READ, it) }
+        isServerAdmin.forTarget<User>()
+            .orElseThrow(UnauthorizedException.All(EntityType.USER, AccessType.READ, currentUser.id.toString()))
+            .checkFor(currentUser, currentUser)
```

Comment:
Could we add a short form when we don't use the target?

This could be
```suggestion
        isServerAdmin
            .orElseThrow(UnauthorizedException.All(EntityType.USER, AccessType.READ, currentUser.id.toString()))
            .checkFor(currentUser)
```

Reply:
I already simplified it a bit, but I agree, maybe I can introduce a single parameter `checkFor`.

Reply:
I think you can add a default value to the second parameter.

Reply:
You cannot add a default value for the second parameter since it is generic. What do you think of overloading the `checkFor` function like this?
```kotlin
suspend fun AccessRule<Unit>.checkFor(requester: User) {
    val allowed = isAllowedToAccess(requester, Unit)
    if (!allowed) throw AccessRuleCheckFailedException()
}
```

Reply:
Good catch @domienderle, I didn't check whether this would work.

We can simply add
```kotlin
suspend fun AccessRule<Unit>.checkFor(requester: User) = checkFor(requester, Unit)
```

Reply:
Ok, then I will add this overloaded function.